### PR TITLE
fix: Fix primary keys format and add property tests to the ORM package

### PR DESCRIPTION
**Pull request by ruhatch**
## Description

Closes: #456

- Fix primary keys format:
  * `bytes` are length prefixed
  * `string` is null-terminated
  * integer types (i.e. `uint32`, `uint64`, etc.) are stored as 4 or 8 big-endian
- Add property tests to the ORM package
- Make `tableBuilder` and `table` private to the `orm` package since there's no way to enforce that the keys used with it follow what is described above (which would result in the same bug as described in https://github.com/regen-network/regen-ledger/issues/456#issue-960703199)

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: orm/primary_key.go [hunk 1]
```diff
@@ -30,10 +32,54 @@ type PrimaryKeyed interface {
...
+func buildPrimaryKey(fields ...interface{}) []byte {
+	bytesSlice := make([][]byte, len(fields))
+	totalLen := 0
+	for i, field := range fields {
+		bytesSlice[i] = primaryKeyFieldBytes(field)
+		totalLen += len(bytesSlice[i])
+	}
+	primaryKey := make([]byte, 0, totalLen)
+	for _, bs := range bytesSlice {
+		primaryKey = append(primaryKey, bs...)
```
Comment (robert-zaremba) [inline orm/primary_key.go:56]: You can use `copy` instead of append to squeeze some performance cycles.

File: orm/primary_key.go [hunk 2]
```diff
@@ -30,10 +32,54 @@ type PrimaryKeyed interface {
...
+	return primaryKey
+}
+
+func primaryKeyFieldBytes(field interface{}) []byte {
+	switch v := field.(type) {
+	case []byte:
+		return AddLengthPrefix(v)
+	case string:
+		// TODO(@ruhatch): Do something with strings
+		return []byte(v)
```
Comment (robert-zaremba) [inline orm/primary_key.go:67]: We have `UnsafeBytesToString` which we should use here to remove unnecessary memory allocations

File: orm/primary_key.go [hunk 3]
```diff
@@ -30,10 +32,54 @@ type PrimaryKeyed interface {
...
+	default:
+		panic(fmt.Sprintf("Type %T not allowed as primary key field", v))
+	}
+}
+
+// TODO(@ruhatch): This isn't super efficient for small cases, should we do something different?
+func AddLengthPrefix(bytes []byte) []byte {
+	prefixedBytes := make([]byte, 0, 8+len(bytes))
+	prefixedBytes = append(prefixedBytes, EncodeSequence(uint64(len(bytes)))...)
+	prefixedBytes = append(prefixedBytes, bytes...)
```
Comment (robert-zaremba) [inline orm/primary_key.go:79]: Use `copy`

File: orm/primary_key.go (hunk 2 above)
Comment (robert-zaremba) [reply orm/primary_key.go:67]: BTW: why we don't Len prefix strings?

Comment (ruhatch) [reply orm/primary_key.go:67]: In the discussion [here](https://github.com/cosmos/cosmos-sdk/discussions/9156), @aaronc suggested using null termination on strings. I haven't implemented it yet, but was planning on doing so if the general approach looked okay.

**Comment by cyberbono3** [comment]
can anyone from west team take care of it since Ru and Marie are on vacation? cc@clevinson

File: orm/sequence_property_test.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+// Init sets up the real Sequence, including choosing a random initial value,
+// and intialises the model state
+func (m *sequenceMachine) Init(t *rapid.T) {
+	// Create context
+	m.ctx = orm.NewMockContext()
+
+	// Create primary key table
+	storeKey := sdk.NewKVStoreKey("test")
+	const testSeqPrefix = iota
```
Comment (technicallyty) [inline orm/sequence_property_test.go:33]:
```suggestion
```

unused variable

File: orm/table.go [hunk 5]
```diff
@@ -188,8 +203,13 @@ func (a Table) Delete(ctx HasKVStore, rowID RowID) error {
 	return nil
 }
 
-// Has checks if a key exists. Panics on nil key.
-func (a Table) Has(ctx HasKVStore, rowID RowID) bool {
+// Has checks if a key exists. Panics on nil or empty key.
```
Comment (technicallyty) [inline orm/table.go:206]: not sure that this comment is true anymore - looks like it will just return false on the empty key

**Review by technicallyty — APPROVED** [review]
looks good, pre-approving

File: orm/primary_key.go [hunk 6]
```diff
@@ -10,43 +12,117 @@ var _ Indexable = &PrimaryKeyTableBuilder{}
...
+// with a length prefix, strings are null-terminated, and integers are encoded
+// using 4 or 8 byte big endian.
+func PrimaryKey(obj PrimaryKeyed) []byte {
+	fields := obj.PrimaryKeyFields()
+	return buildPrimaryKey(fields...)
+}
+
+// buildPrimaryKey encodes and concatenates the PrimaryKeyFields. See PrimaryKey
+// for full documentation of the encoding.
+func buildPrimaryKey(fields ...interface{}) []byte {
```
Comment (robert-zaremba) [inline orm/primary_key.go]: I would prefer that this function and all related functions return error rather than panic.

File: orm/primary_key.go [hunk 7]
```diff
@@ -10,43 +12,117 @@ var _ Indexable = &PrimaryKeyTableBuilder{}
...
+
+// Prefix the byte array with its length as 8 bytes
+func AddLengthPrefix(bytes []byte) []byte {
+	byteLen := len(bytes)
+	if byteLen > 256 {
+		panic("Cannot create primary key with an []byte of length greater than 256 bytes. Try again with a smaller []byte.")
+	}
+
+	newLen := 1 + len(bytes)
+	prefixedBytes := make([]byte, newLen, newLen)
```
Comment (robert-zaremba) [inline orm/primary_key.go]: `gosimple` says that we don't need to put cap

File: orm/table.go [hunk 8]
```diff
@@ -188,8 +203,13 @@ func (a Table) Delete(ctx HasKVStore, rowID RowID) error {
...
+func (a table) Has(ctx HasKVStore, rowID RowID) bool {
```
Comment (robert-zaremba) [inline orm/table.go]: I will rename `rowID` to `key` to match the code description / comment

File: orm/table.go (hunk 5 above)
Comment (robert-zaremba) [reply orm/table.go]: True - we accept empty key, but panic on `nil`.

**Review by robert-zaremba — APPROVED** [review]
tACK. Added tests and pushed cosmetic changes and comment updates. We should note when a function panics.

File: orm/primary_key.go (hunk 6 above)
Comment (robert-zaremba) [reply orm/primary_key.go]: I've updated the comment, and left function signature, but will be worth to consider not panicing.

File: orm/sequence_property_test.go (hunk 4 above)
Comment (clevinson) [reply orm/sequence_property_test.go]: @robert-zaremba you resolved- but did you actually commit these suggestions? Or is there some reason this `testSeqPrefix` is still necessary?

Comment (technicallyty) [reply orm/sequence_property_test.go]: it looks like its used now

File: orm/table.go (hunk 5 above)
Comment (technicallyty) [reply orm/table.go]: the nil value here would just be the empty []byte slice though, so there shouldn't ever be a panic -->> https://play.golang.org/p/eLzItdNL8He

File: orm/sequence_property_test.go (hunk 4 above)
Comment (robert-zaremba) [reply orm/sequence_property_test.go]: I included it in other changes: https://github.com/regen-network/regen-ledger/pull/459/commits/1075cee0b52bfcbca428c831983944c0d592f370

File: orm/table.go (hunk 5 above)
Comment (robert-zaremba) [reply orm/table.go]: right, will do a PR to update the comment.

**Pull request merged**